contracts/meta/MetaSwapUtils.sol

```diff
@@ -0,0 +1,1184 @@
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+    using MathUtils for uint256;
+    using AmplificationUtils for SwapUtils.Swap;
+
+    /*** EVENTS ***/
+
+    event TokenSwap(
+        address indexed buyer,
+        uint256 tokensSold,
+        uint256 tokensBought,
+        uint128 soldId,
+        uint128 boughtId
+    );
+    event TokenSwapUnderlying(
+        address indexed buyer,
+        uint256 tokensSold,
+        uint256 tokensBought,
+        uint128 soldId,
+        uint128 boughtId
+    );
+    event AddLiquidity(
+        address indexed provider,
+        uint256[] tokenAmounts,
+        uint256[] fees,
+        uint256 invariant,
+        uint256 lpTokenSupply
+    );
+    event RemoveLiquidityOne(
+        address indexed provider,
+        uint256 lpTokenAmount,
+        uint256 lpTokenSupply,
+        uint256 boughtId,
+        uint256 tokensBought
+    );
+    event RemoveLiquidityImbalance(
+        address indexed provider,
+        uint256[] tokenAmounts,
+        uint256[] fees,
+        uint256 invariant,
+        uint256 lpTokenSupply
+    );
+    event NewAdminFee(uint256 newAdminFee);
+    event NewSwapFee(uint256 newSwapFee);
+    event NewWithdrawFee(uint256 newWithdrawFee);
+
+    struct MetaSwap {
+        // Meta-Swap related parameters
+        ISwap baseSwap;
+        uint256 baseVirtualPrice;
+        uint256 baseCacheLastUpdated;
+        IERC20[] baseTokens;
+    }
+
+    // Struct storing variables used in calculations in the
+    // calculateWithdrawOneTokenDY function to avoid stack too deep errors
+    struct CalculateWithdrawOneTokenDYInfo {
+        uint256 d0;
+        uint256 d1;
+        uint256 newY;
+        uint256 feePerToken;
+        uint256 preciseA;
+        uint256 xpi;
+    }
+
+    // Struct storing variables used in calculation in removeLiquidityImbalance function
+    // to avoid stack too deep error
+    struct ManageLiquidityInfo {
+        uint256 d0;
+        uint256 d1;
+        uint256 d2;
+        LPToken lpToken;
+        uint256 totalSupply;
+        uint256 preciseA;
+        uint256 baseVirtualPrice;
+        uint256[] tokenPrecisionMultipliers;
+        uint256[] newBalances;
+    }
+
+    struct SwapUnderlyingInfo {
+        uint256 x;
+        uint256 dx;
+        uint256 dy;
+        uint8 metaIndexFrom;
+        uint8 metaIndexTo;
+        uint256[] oldBalances;
+        IERC20[] baseTokens;
+        IERC20 tokenFrom;
+        IERC20 tokenTo;
+        uint256 baseVirtualPrice;
+    }
+
+    struct CalculateSwapUnderlyingInfo {
+        uint256 baseVirtualPrice;
+        ISwap baseSwap;
+        uint8 baseLPTokenIndex;
+        uint8 baseTokensLength;
+        uint8 metaIndexTo;
+        uint256 x;
+        uint256 dy;
+    }
+
+    // the denominator used to calculate admin and LP fees. For example, an
+    // LP fee might be something like tradeAmount.mul(fee).div(FEE_DENOMINATOR)
+    uint256 private constant FEE_DENOMINATOR = 10**10;
```

Comment:
Just curious, why not 10**9 for 1 gwei? lol

Reply:
No particular reason - though it would probably make more sense to use 10**9 or bps unit.